Data-Structures/Utility-Collections.md

```diff
@@ -0,0 +1,260 @@
+
+## Collection types
+
+Collections can be broadly categorized across two dimensions: singular or associative and ordered or unordered:
+
+ * **Singular** collections that contain stand-alone entities. These entities may be complex but do not depend on an association with a "key".
+ * **Associative** collections that contain pairs, in the form of a "key" and a "value". Where the "key" is generally a simple type, such as a string, and the "value" can be complex.
+
+_The delineation between singular and associative collections is not strict. An indexed vector or array could be categorized as a key/value store where the keys are the indexes, and a Map whose values are ignored could be used as a Set._
+
+ * **Ordered** collections expose their elements in some form of ordering. Ordering can either be "natural" based on some comparator of their elements or keys, or be dictated by construction, such as a vector whose elements are indexed by the order in which they are added. Iterators on ordered collections become useful for search operations, particularly when subset-selection is possible (i.e. range queries).
+ * **Unordered** collections do not expose their elements in any useful order. Unordered collections are generally more efficient in their layout and access for large amounts of data and their algorithms are often simpler than that for ordered collections.
+
+Below is a list of collection types that have been found useful across many programming platforms. This currently represents a _wish-list_ for IPLD and pointers to areas of research. It is anticipated that multiple forms of any one type of IPLD collection may emerge, using different schemas and algorithms and being suited to different types and size of data.
+
+These collections are not necessarily distinct. It is likely that IPLD collection implementations can serve as multiple collection types among this list. For example: a HashMap can also act as a HashSet (in fact it is common for implementations of these to share underlying components), or a Vector could expose operations that allow it to act as a simple Queue or Stack. It is probable that a small number of algorithms will serve as primitives for building many, or all of these data structures. Users should be able to select the interface that meet their programming needs and the implementations that best serve their data type and size needs. Trade-offs will be common and users will need to be presented with sufficient information to make appropriate choices.
+
+_When listing collection operations below, those indicated with a '`?`' are not strictly necessary for a useful implementation of that collection type but are commonly made available for convenience._
+
+### Singular collections
+
+#### Unordered
+
+##### `Set`
+
+A simple collection of non-duplicate entries stored without useful ordering.
+
+Operations include:
+
+* `Add(element)`
+* `Remove(element)`
+* `Iterate()`
+* `Contains(element)?`
+* `Size()?`
```

Comment:
I’m always concerned when we encode information like this into the root node, because you could just lie ;)

Without traversing the entire structure you don’t actually know the real size. For instance, we encode the length of files into `unixfs-v2` but it can’t really be trusted. It’s useful in interfaces for showing “this file/directory is {size}” without actually requesting all the blocks, but I’m constantly worried people will trust it when they shouldn’t.

If we put a size property in the base interface I’m concerned that it will lead people to trust it, and it’s actually very easy to lie.

Reply:
I don't think that's built into this document though, I'm leaving that open to the implementation and just listing this as a possible _operation_. In fact, I was just assuming here that it would be implemented as an actual traversal calculation if it's implemented at all. It's what I did in [iamap](https://github.com/rvagg/iamap/blob/bfa8c77e798f9ce4f6e9e408458b51c29002eb12/iamap.js#L428-L435) but the user has to know the cost of calling it!